### PR TITLE
Generate cross platform headers

### PR DESCRIPTION
**Pull request by mars979**
We previously used CMake's generate_export_header to generate platform dependent export header. This will allow to generate one in all cross platform header for Win, Linux, MacOS and Android

### CODE REVIEW

**Comment by JeroMiya** [comment]
Were you able to test this change on OSVR-Android-Build?

**Comment by mars979** [comment]
@JeroMiya , I checked the headers from the Android build and they were the same as linux ones so it should build there as well

**Comment by JeroMiya** [comment]
Cool. Go ahead and merge this when you're ready.

File: src/osvr/Macros.cmake [hunk 1]
```diff
@@ -34,10 +39,19 @@ macro(osvr_add_library)
 
     set_target_properties(${LIBNAME_FULL} PROPERTIES
         FOLDER "${LIB_FOLDER}")
-
-    generate_export_header(${LIBNAME_FULL}
-        BASE_NAME ${EXPORT_BASENAME}
-        EXPORT_FILE_NAME Export.h)
+        
+    # Are we building or using this library?
+    get_target_property(EXPORT_IMPORT_CONDITION ${LIBNAME_FULL} DEFINE_SYMBOL)
```
Comment (rpavlik) [inline src/osvr/Macros.cmake]: I don't follow the purpose of this section of code.  CMake auto-defines some preprocessor define `${LIBNAME_FULL}_EXPORTS`

File: src/osvr/exportheader.cmake.in [hunk 2]
```diff
@@ -0,0 +1,89 @@
+
```
Comment (rpavlik) [inline src/osvr/exportheader.cmake.in]: Needs file header with license, etc - copy from another one, make sure to note that it's a generated file.

File: src/osvr/Macros.cmake (hunk 1 above)
Comment (rpavlik) [inline src/osvr/Macros.cmake]: yeah, between 42 and 50 doesn't appear to actually even be used at all, and can probably be removed.

File: src/osvr/Macros.cmake [hunk 3]
```diff
@@ -34,10 +39,10 @@ macro(osvr_add_library)
 
     set_target_properties(${LIBNAME_FULL} PROPERTIES
         FOLDER "${LIB_FOLDER}")
-
-    generate_export_header(${LIBNAME_FULL}
-        BASE_NAME ${EXPORT_BASENAME}
-        EXPORT_FILE_NAME Export.h)
+
```
Comment (rpavlik) [inline src/osvr/Macros.cmake]: You aren't using an editor that removes whitespace at the end of lines?

File: src/osvr/exportheader.cmake.in [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      /* We are building this library */
+#    define @EXPORT_BASENAME@_EXPORT __attribute__((dllexport))
+#  else
+      /* We are using this library */
+#    define @EXPORT_BASENAME@_EXPORT __attribute__((dllimport))
+#  endif
+#endif
+
+#ifndef @EXPORT_BASENAME@_NO_EXPORT
+#  define @EXPORT_BASENAME@_NO_EXPORT __attribute__((visibility("hidden")))
```
Comment (rpavlik) [inline src/osvr/exportheader.cmake.in]: I don't think you can do `__attribute__((visibility("hidden")))` on Windows, can you? Do you want me to try building this branch on msys/mingw?

Comment (rpavlik) [inline src/osvr/exportheader.cmake.in]: that is to say, I believe it defaults to hidden, so _NO_EXPORT would just be an empty define

File: src/osvr/exportheader.cmake.in [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+#endif
+
+#ifndef @EXPORT_BASENAME@_DEPRECATED
+#  define @EXPORT_BASENAME@_DEPRECATED __attribute__ ((__deprecated__))
+#endif
+
+#elif (defined(_GNUC_) && _GNUC_C >= 4) || defined(__APPLE__) ||               \
```
Comment (rpavlik) [inline src/osvr/exportheader.cmake.in]: Mac OS X is a *nix technically, and android kind of is too.  Nitpicking, but lowercase the n and put an X in for Mac OS X

File: src/osvr/exportheader.cmake.in [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+#ifndef @EXPORT_BASENAME@_DEPRECATED_EXPORT
+#  define @EXPORT_BASENAME@_DEPRECATED_EXPORT @EXPORT_BASENAME@_EXPORT @EXPORT_BASENAME@_DEPRECATED
+#endif
+
+/* fallback def */
+#ifndef @EXPORT_BASENAME@_DEPRECATED_NO_EXPORT
+#  define @EXPORT_BASENAME@_DEPRECATED_NO_EXPORT @EXPORT_BASENAME@_NO_EXPORT @EXPORT_BASENAME@_DEPRECATED
+#endif
+
+#endif
```
Comment (rpavlik) [inline src/osvr/exportheader.cmake.in]: Missing newline at end of file!

File: src/osvr/exportheader.cmake.in (hunk 2 above)
Comment (mars979) [inline src/osvr/exportheader.cmake.in]: Added license and a note that it's a generated file

File: src/osvr/Macros.cmake (hunk 1 above)
Comment (mars979) [inline src/osvr/Macros.cmake]: I think you are right. Works without it. I removed unused portion

File: src/osvr/Macros.cmake (hunk 3 above)
Comment (mars979) [inline src/osvr/Macros.cmake]: hehe, nope, I tend to use notepad++ for CMake scripts and it doesn't trim trailing whitespace automatically (there is a macro though). I removed trailing spaces

File: src/osvr/exportheader.cmake.in (hunk 4 above)
Comment (mars979) [inline src/osvr/exportheader.cmake.in]: yes, if you could try a build, that'd be great

Comment (mars979) [inline src/osvr/exportheader.cmake.in]: Fixed

**Comment by mars979** [comment]
@rpavlik , pushed changes according to your comments. If it looks good to you, please merge it when you can

**Comment by rpavlik** [comment]
OK, I pulled this locally and tested it out.  Found a few small issues, which I fixed, and did some enhancements based on some of the logic found in the CMake module here: https://github.com/Kitware/CMake/blob/master/Modules/GenerateExportHeader.cmake#L206

I tested locally by building with MSYS2/MinGW (so that's GCC on Windows) which seemed OK. Will test here with MSVC2013 in a moment. If that works and Travis works, then we should be good to merge.

**Comment by rpavlik** [comment]
OK, travis and windows are good, :+1:

**Pull request merged**